test/fluree/db/shacl/shacl_basic_test.clj

```diff
@@ -1477,3 +1477,76 @@
       (is (util/exception? invalid-hand))
       (is (= "SHACL PropertyShape exception - path [[1003 :predicate]] conformed to sh:qualifiedValueShape fewer than sh:qualifiedMinCount times."
              (ex-message invalid-hand))))))
+
+(deftest ^:integration post-processing-validation
+  (let [conn @(fluree/connect {:method :memory})
+        ledger @(fluree/create conn "post-processing" {:defaultContext [test-utils/default-str-context
+                                                                        {"ex" "http://example.com/"}]})
+        db0 (fluree/db ledger)]
+    (testing "shacl-objects-of-test"
+      (let [db1 @(fluree/stage db0
+                               [{"@id" "ex:friendShape"
+                                 "type" ["sh:NodeShape"]
+                                 "sh:targetObjectsOf" {"@id" "ex:friend"}
+                                 "sh:property" [{"sh:path" {"@id" "ex:name"}
+                                                 "sh:datatype" {"@id" "xsd:string"}}]}])
+            db2 @(fluree/stage db1 [{"id" "ex:Bob"
+                                     "ex:name" 123
+                                     "type" "ex:User"}])
+            db-forbidden-friend @(fluree/stage db2
+                                               {"id" "ex:Alice"
+                                                "type" "ex:User"
+                                                "ex:friend" {"@id" "ex:Bob"}})]
+        (is (util/exception? db-forbidden-friend))
+        (is (= "SHACL PropertyShape exception - sh:datatype: every datatype must be 1."
```

Comment:
I don't think this matches the error message you'd get with the mid-transaction-processing version of this constraint check. If we're going to keep those versions around, I think the error messages should match -- from the user's perspective, it's the same constraint, so different error messages for post/mid processing wouldn't make sense.

Reply:
This is an interesting one, when I try a naive mid-processing datatype validation I actually get a coercion error instead of a validation error, which makes sense as coercion happens before validation.